### ISSUE TITLE
questions on multi-thread traning

### ISSUE THREAD

**Issue by ecolss** [aksnzhy/xlearn#92]
As said in the doc, xlearn could do Hogwild! or lock-free training, and as I read the code, so for example,
`CrossEntropyLoss::CalcGrad` splits the samples into parts and feeds them into multiple threads,
then each thread will call, say `FFMScore::CalcGrad` to calculate the gradients and update the weights.

My dumb question is that, since each thread update weights simultaneously, and it's possible 2 threads may try to update same feature's weight vector at the same time (I know the basic idea of Hogwild!), but will this cause False Sharing issue which slows down the performance?

For example, I have a dataset with 1m rows, each row has a feature `country`, and there are 4 unique `country` values, when doing training in`xlearn`, one batch (corresponding `reader->Samples` in `xlearn`) is highly possible to have different rows with same `country` values, and they split to different threads, and these threads will try to update the feature weight vector of this same `country` at the same time.

**Comment by aksnzhy** [comment]
Yes, the lock free training can improve CPU performance but may hurt convergence. This is a trade-off. If your data is sparse, lock free training often performe well, but if your data is dense, I suggest that you can turn off the lock free flag.
